// @amuzlab/ftp-client/index.js
'use strict'

const FtpClient = require('./FtpClient')

Object.defineProperties(
    exports,
    {
        /**
         * @name        FtpClient
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @readonly
         * @see         {@link ftp-client~FtpClient FtpClient}
         * @type        {ftp-client~FtpClient}
        */
        FtpClient: {
            enumerable: true,
            value: FtpClient
        },
        /**
         * @name        ERROR
         * @since       0.0.1
         * @author      예장해
         * @public
         * @static
         * @readonly
         * @see         {@link module:error error}
         * @type        {module:error}
        */
        ERROR: {
            enumerable: true,
            value: require('./error')
        },
        /**
         * @name        SOURCE_TYPE
         * @since       0.0.1
         * @author      예장해
         * @public
         * @static
         * @readonly
         * @see         {@link module:sourceType sourceType}
         * @type        {module:sourceType}
        */
        SOURCE_TYPE: {
            enumerable: true,
            value: require('./sourceType')
        },
        /**
         * @name        STATUS
         * @since       0.0.1
         * @author      예장해
         * @public
         * @static
         * @readonly
         * @see         {@link module:status status}
         * @type        {module:status}
        */
        STATUS: {
            enumerable: true,
            value: require('./status')
        },
        /**
         * @name        validation
         * @since       0.0.1
         * @author      예장해
         * @public
         * @static
         * @readonly
         * @see         {@link module:validation validation}
         * @type        {module:validation}
        */
        validation: {
            enumerable: true,
            value: require('./validation')
        },
        /**
         * @name        createFtpClient
         * @function
         * @desc        전달된 설정 데이터로 FtpClient instance를 생성하여 리턴
         * @since       0.0.1
         * @author      예장해
         * @public
         * @static
         * @see         {@link ftp-client~FtpClient FtpClient}
         * @param       {Object}    config  - {@link ftp-client~FtpClient FTP 설정 데이터}
         * @return      {ftp-client~FtpClient}  FtpClient instance
        */
        createFtpClient: {
            enumerable: true,
            value: config => {
                return new FtpClient(config)
            }
        }
    })
