// @amuzlab/ftp-client/status.js
'use strict'

/**
 * @since       0.0.1
 * @module      status
 * @desc        ftp-client의 상태를 정의
 * @author      예장해
 * @date        2019-01-23
*/

Object.defineProperties(
    exports,
    {
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   CONNECTED
         * @default     0
         * @desc        FTP 서버와 연결된 상태
        */
        CONNECTED: {
            enumerable: true,
            value: 0
        },
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   CONNECTING
         * @default     1
         * @desc        FTP 서버와 연결중인 상태
        */
        CONNECTING: {
            enumerable: true,
            value: 1
        },
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   DISCONNECTED
         * @default     2
         * @desc        FTP 서버와 연결이 끊어진 상태
        */
        DISCONNECTED: {
            enumerable: true,
            value: 2
        },
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   DISCONNECTING
         * @default     3
         * @desc        FTP 서버와 연결을 끊고 있는 상태
        */
        DISCONNECTING: {
            enumerable: true,
            value: 3
        },
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   ERROR
         * @default     4
         * @desc        에러 상태
        */
        ERROR: {
            enumerable: true,
            value: 4
        },
        /**
         * @since       0.0.1
         * @author      예장해
         * @date        2019-01-23
         * @public
         * @static
         * @constant    {Integer}   RECONNECTING
         * @default     5
         * @desc        FTP 서버로 재접속 시도중인 상태
        */
        RECONNECTING: {
            enumerable: true,
            value: 5
        }
    })
